src/org/zaproxy/zap/extension/websocket/WebSocketEventPublisher.java

```diff
@@ -0,0 +1,154 @@
+
+import java.net.Socket;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+
+import org.zaproxy.zap.ZAP;
+import org.zaproxy.zap.eventBus.Event;
+import org.zaproxy.zap.eventBus.EventPublisher;
+import org.zaproxy.zap.extension.websocket.WebSocketProxy.Initiator;
+import org.zaproxy.zap.extension.websocket.WebSocketProxy.State;
+
+/**
+ * A {@link EventPublisher} of websocket events.
+ * @since TODO add version
+ */
+public final class WebSocketEventPublisher implements EventPublisher, WebSocketSenderListener {
+
+    /**
+     * The event sent when a {@code WebSocketMessage} is seen.
+     */
+    public static final String EVENT_MESSAGE = "ws.message";
+    /**
+     * The event sent when a {@code WebSocketProxy} state change occurs.
+     */
+    public static final String EVENT_STATE_CHANGE = "ws.stateChange";
+
+    private static final String FIELD_STATE = "state";
+    private static final String FIELD_CHANNEL_ID = "channelId";
+    private static final String FIELD_LOCAL_SOCKET = "localSocket";
+    private static final String FIELD_REMOTE_SOCKET = "remoteSocket";
+    private static final String FIELD_TIME_IN_MS = "timeSentInMs";
+    private static final String FIELD_OP_CODE = "opCode";
+    private static final String FIELD_OP_CODE_STRING = "opCodeString";
+    private static final String FIELD_DIRECTION = "direction";
+    private static final String FIELD_LENGTH = "length";
+    private static final String FIELD_MSG_ID = "messageId";
+    private static final String FIELD_MSG_SUMMARY = "messageSummary";
+
+    private ExtensionWebSocket extension;
+    private ExecutorService executor;
+
+    public WebSocketEventPublisher (ExtensionWebSocket extension) {
+        this.extension = extension;
+        executor = Executors.newSingleThreadExecutor(new ThreadFactory() {
+
+            @Override
+            public Thread newThread(Runnable runnable) {
+                return new Thread(runnable, "ZAP-WebSocketEventPublisher");
+            }});
+        
+        ZAP.getEventBus().registerPublisher(
+                this,
+                new String [] {EVENT_MESSAGE, EVENT_STATE_CHANGE});
+    }
+    
+    @Override
+    public String getPublisherName() {
+        return WebSocketEventPublisher.class.getCanonicalName();
+    }
+
+    @Override
+    public int getListenerOrder() {
+        // We want this to run after most listeners
+        return 100000;
+    }
+
+    @Override
+    public void onMessageFrame(int channelId, WebSocketMessage message, Initiator initiator) {
+        WebSocketProxy proxy = extension.getWebSocketProxy(channelId);
+        if (proxy.isAllowAPI()) {
+            // Sending an event on an API message will cause an infinite loop
+            return;
+        }
+        if (message.isFinished) {
+            Map<String, String> map = new HashMap<String, String>();
```

Comment:
The event creation could be done in the other thread. (Same commend for the other event)